SearchFilter to support JSONField and HStoreField

hishnash opened the pull request:
Since JSONFields and HStoreFields use `__` we needed to update the m2m checking code to handle search_fields that contain __ that are not relationships.

## Description

For `SearchFilter` to support [JSONFields](https://docs.djangoproject.com/en/3.0/ref/contrib/postgres/fields/#querying-jsonfield) and [HStoreFields](https://docs.djangoproject.com/en/3.0/ref/contrib/postgres/fields/#querying-hstorefield).

This enables:
`search_fields = ['jsonfieldname__nestedkey__path',]`.

### The Bug:

When traversing the `__` separated search_field path the code expects the `opts` value to be updated on each iteration of the traversal. However the JsonField/HStoreField does not provide a `get_path_info` (since it is not a relation). This results in the next iteration of the traversal reusing the old `opts` value and raising an error (`MyModel has not attribute nestedkey`). 

### The fix:

When we check if it is a `relation` if it is not we can stop traversing that search_field and continue on to the next search_field.

rpkilby commented:
Hi @hishnash. Like #6924, tests to verify the result will be required. My recommendation would be to test against a mock queryset, and assert that the appropriate filter calls are provided. This will avoid running a query against the DB.

hishnash commented:
@rpkilby do you recon tests that use JSONField and HStoreField or just test that validate that SearchFilter supports fields with custom Transforms (__ operators).

hishnash commented:
@rpkilby I have added a test for this.

since this change does not just apply to the Postgres only operators but also to any other custom transforms.

I added a custom `TrimA` transform during the test this way we don't need to make the tests depend on Postgres. 

Do you think we should add something to the docs with respect to JSONField/HStoreField ?

hishnash commented:
Is there anything I can do here to help this with its progress to be merged?

tomchristie commented:
Code wise it looks neatly done, yup.

I think it'd probably need a note in an appropriate place in the docs in order for anyone to figure out that it's there, and how to make use of it. (Tho I've not had a look to figure out where would be a good place to document it?)

hishnash commented:
@tomchristie  thank you I will take a look add some proposed docs to this PR.

hishnash commented:
@tomchristie I have added some documentation related to JSON and HStore fields. 

Im not sure we should talk about custom `transforms` in the docs, it might confuse non-advanced users.  If you are a user using them I believe you would just try to do `__mytransfrom` (as you would normally in django) and find that it works.

tomchristie commented:
Yup, looks like a good addition - thanks!
